workflow/controller/operator.go

```diff
@@ -2311,34 +2311,36 @@ func getTemplateOutputsFromScope(tmpl *wfv1.Template, scope *wfScope) (*wfv1.Out
 
 // hasOutputResultRef will check given template output has any reference
 func hasOutputResultRef(name string, parentTmpl *wfv1.Template) bool {
-	var varRefNamePattern string
+	varRefNamePattern := fmt.Sprintf(`\.%s\.outputs\.result`, name)
 	if parentTmpl.DAG != nil {
-		varRefNamePattern = "tasks([[.](['\"])?)" + name + "((['\"])?]?).outputs.result"
+		varRefNamePattern = "tasks" + varRefNamePattern
 	} else if parentTmpl.Steps != nil {
-		varRefNamePattern = "steps([[.](['\"])?)" + name + "((['\"])?]?).outputs.result"
+		varRefNamePattern = "steps" + varRefNamePattern
 	}
 
 	jsonValue, err := json.Marshal(parentTmpl)
 	if err != nil {
 		log.Warnf("Unable to marshal the template. %v, %v", parentTmpl, err)
 	}
 
-	contain, err := regexp.MatchString(varRefNamePattern, string(jsonValue))
+	contains, err := regexp.MatchString(varRefNamePattern, string(jsonValue))
 	if err != nil {
-		log.Warnf("Error in Regex compilation. %s, %v", varRefNamePattern, err)
+		log.Warnf("Error in regex compilation %q: %v", varRefNamePattern, err)
 	}
-	return contain
+	return contains
 }
 
 // getStepOrDAGTaskName will extract the node from NodeStatus Name
 func getStepOrDAGTaskName(nodeName string) string {
-	if strings.Contains(nodeName, ".") {
-		name := nodeName[strings.LastIndex(nodeName, ".")+1:]
-		// Retry, withItems and withParam scenario
-		if indx := strings.Index(name, "("); indx > 0 {
-			return name[0:indx]
-		}
-		return name
+	// If our name contains an open parenthesis, this node is a child of a Retry node or an expanded node
+	// (e.g. withItems, withParams, etc.). Ignore anything after the parenthesis.
+	if parenthesisIndex := strings.Index(nodeName, "("); parenthesisIndex >= 0 {
+		nodeName = nodeName[:parenthesisIndex]
+	}
+	// If our node contains a dot, we're a child node. We're only interested in the step that called us, so return the
+	// name of the node after the last dot.
+	if lastDotIndex := strings.LastIndex(nodeName, "."); lastDotIndex >= 0 {
+		nodeName = nodeName[lastDotIndex+1:]
```

Comment:
Previously this functioned assumed that the only way to have a `.` in a node name is if a node is a child of another node. This is not true. A node name could have a `.` from an expanded item: e.g.

```
"data-transformation-gjrt8[0].generate-artifact(2:foo/script.py)"
                             ^ Expected dot in name         ^ Valid (user-provided) dot.
                                                              Should not be considered in function.
```

Similarly for `(`